crates/wasi/src/preview2/poll.rs

```diff
@@ -1,174 +1,257 @@
 use crate::preview2::{bindings::io::poll, WasiView};
-use anyhow::Result;
-use std::any::Any;
-use std::collections::HashMap;
-use std::future::Future;
-use std::pin::Pin;
-use std::task::{Context, Poll};
-use wasmtime::component::{Resource, ResourceTable};
-
-pub type PollableFuture<'a> = Pin<Box<dyn Future<Output = ()> + Send + 'a>>;
-pub type MakeFuture = for<'a> fn(&'a mut dyn Any) -> PollableFuture<'a>;
-pub type ClosureFuture = Box<dyn Fn() -> PollableFuture<'static> + Send + 'static>;
+use anyhow::{anyhow, Result};
+use futures::Future;
+use smallvec::{smallvec, SmallVec};
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    pin::Pin,
+    task::{Context, Poll},
+};
+use wasmtime::component::{Lease, Resource, ResourceTable};
 
 /// A host representation of the `wasi:io/poll.pollable` resource.
 ///
-/// A pollable is not the same thing as a Rust Future: the same pollable may be used to
-/// repeatedly check for readiness of a given condition, e.g. if a stream is readable
-/// or writable. So, rather than containing a Future, which can only become Ready once, a
-/// Pollable contains a way to create a Future in each call to `poll`.
-pub struct Pollable {
-    index: u32,
-    make_future: MakeFuture,
-    remove_index_on_delete: Option<fn(&mut ResourceTable, u32) -> Result<()>>,
+/// A pollable is not the same thing as a Rust Future: in WASI, the same pollable
+/// may be used to repeatedly check for readiness of a given condition, e.g. if
+/// a stream is readable or writable.
+pub trait Pollable: Send + 'static {
+    /// Check the current readiness status. When the pollable is not ready yet,
+    /// `poll_ready` returns [Poll::Pending] and stores a clone of the Waker
+    /// copied from the current [Context]. Unlike [std::future::Future::poll],
+    /// this method _will_ be called again after it previously returned
+    /// [Poll::Ready]. This is normal behavior and should not panic.
+    fn poll_ready(&mut self, cx: &mut Context<'_>, view: &mut dyn WasiView) -> Poll<()>;
 }
 
+/// Convenience trait for implementing [Pollable] in terms of an `async` method.
+/// There is a blanket implementation of [Pollable] for all [Subscribe]'s,
+/// so all Subscribe implementations are automatically Pollable.
 #[async_trait::async_trait]
 pub trait Subscribe: Send + 'static {
+    /// Wait for the pollable to be ready.
+    ///
+    /// # Cancel safety
+    /// The implementation must make sure to only await futures that are
+    /// cancel-safe, as the returned future will most liekly be canceled, even
+    /// during normal operation.
     async fn ready(&mut self);
 }
 
-/// Creates a `pollable` resource which is susbcribed to the provided
-/// `resource`.
+impl<T: Subscribe> Pollable for T {
+    fn poll_ready(&mut self, cx: &mut Context<'_>, _view: &mut dyn WasiView) -> Poll<()> {
+        self.ready().as_mut().poll(cx)
+    }
+}
```

Comment:
This I think is actually subtly incorrect because it drops the future after the call to `poll` which signals that the future should be cancelled rather than keeping it alive until the whole poll is done. I think that means that we may not be guaranteed to get wakeups from cancelled futures, although we might get those for now given how the code is currently constructed. I think that this'll need to be a bit fancier with the adapter instead of having a blanket impl

Reply:
The tests were failing precisely because of that. To fix it, I ended up rougly back at the `Subscribe` design, but this time returning a custom WasiFuture type that has an additional `WasiView` parameter on its `poll` method. To prevent scope creep of this PR, I kept `Subscribe` alive for now (as `PollableAsync`). But in the long run, I don't think there's much value in havin them both, as `Subscribe` can be trivially converted to `Pollable` from:

```rs
#[async_trait::async_trait]
impl PollableAsync for HostFutureIncomingResponse {
    async fn ready(&mut self) {
        if let Self::Pending(handle) = self {
            *self = Self::Ready(handle.await);
        }
    }
}
```

to:

```rs
impl Pollable for HostFutureIncomingResponse {
    fn ready<'a>(&'a mut self) -> Pin<Box<dyn WasiFuture<Output = ()> + Send + 'a>> {
        Box::pin(async {
            if let Self::Pending(handle) = self {
                *self = Self::Ready(handle.await);
            }
        })
    }
}
```

One obvious downside is the added visual noise.